src/algorithms/StatePreparation.cpp

```diff
@@ -0,0 +1,285 @@
+/*
+ * Copyright (c) 2025 Chair for Design Automation, TUM
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier: MIT
+ *
+ * Licensed under the MIT License
+ */
+
+#include "algorithms/StatePreparation.hpp"
+
+#include "Definitions.hpp"
+#include "circuit_optimizer/CircuitOptimizer.hpp"
+#include "ir/QuantumComputation.hpp"
+#include "ir/operations/OpType.hpp"
+#include "ir/operations/Operation.hpp"
+#include "ir/operations/StandardOperation.hpp"
+
+#include <algorithm>
+#include <cmath>
+#include <complex>
+#include <cstddef>
+#include <cstdint>
+#include <iterator>
+#include <numeric>
+#include <tuple>
+#include <vector>
+
+static const double EPS = 1e-10;
+
+namespace qc {
+using Matrix = std::vector<std::vector<double>>;
+
+template <typename T> [[nodiscard]] auto twoNorm(std::vector<T> vec) -> double {
+  double norm = 0;
+  for (auto elem : vec) {
+    norm += std::norm(elem);
+  }
+  return sqrt(norm);
+}
+
+template <typename T>
+[[nodiscard]] auto isNormalized(std::vector<T> vec) -> bool {
+  return std::abs(1 - twoNorm(vec)) < EPS;
+}
+
+[[nodiscard]] auto kroneckerProduct(Matrix matrixA, Matrix matrixB) -> Matrix {
```

Comment:
A nd B should be passed as const references here to avoid copying them on every invocation